cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -0,0 +1,168 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.store;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractDataset;
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scan;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.common.utils.ImmutablePair;
+import co.cask.cdap.data2.dataset2.lib.table.FuzzyRowFilter;
+import co.cask.cdap.proto.Id;
+import com.google.common.collect.ImmutableList;
+import com.google.common.primitives.Longs;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.inject.Inject;
+import org.apache.avro.generic.GenericData;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ *
+ */
+public class WorkflowDataset extends AbstractDataset {
+
+  private static final Gson GSON = new Gson();
+  private static final String RUNID = "runId";
+  private static final String TIME_TAKEN = "timeTaken";
+  private static final String NODES = "nodes";
+  private static final String DELIMITER = ":";
+  private static final Logger LOG = LoggerFactory.getLogger(WorkflowDataset.class);
+
+  private final Table table;
+
+  @Inject
+  public WorkflowDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  private byte[] createRowKey(String key, long time) {
+    byte[] result = Bytes.toBytes(key);
+    byte[] rowKey = new byte[result.length + 8];
+    int offset = Bytes.putBytes(rowKey, 0, result, 0, result.length);
+    Bytes.putBytes(rowKey, offset, Bytes.toBytes(time), 0, Bytes.SIZEOF_LONG);
+    return rowKey;
+  }
+
+  public void write(Id.Program id, RunRecordMeta runRecordMeta, List<ActionRuns> actionRunsList) {
+
+    long start = runRecordMeta.getStartTs();
+    String key = id.getApplication().getNamespaceId() + DELIMITER
+      + id.getApplicationId() + DELIMITER + id.getId() + DELIMITER;
+    byte[] rowKey = createRowKey(key, start);
+
+    long timeTaken = runRecordMeta.getStopTs() - start;
+
+    String value = GSON.toJson(actionRunsList, new TypeToken<List<ActionRuns>>() { }.getType());
+
+    table.put(rowKey, Bytes.toBytes(RUNID), Bytes.toBytes(runRecordMeta.getPid()));
+    table.put(rowKey, Bytes.toBytes(TIME_TAKEN), Bytes.toBytes(timeTaken));
+    table.put(rowKey, Bytes.toBytes(NODES), Bytes.toBytes(value));
+  }
+
+  public List<WorkflowRunRecord> scan(Id.Program id, long timeRangeStart, long timeRangeEnd) {
```

Comment:
If this is purely for workflows, lets accept ``Id.Workflow`` instead of ``Id.Program``